content/blog/purrr-walk-this-way/index.Rmd

```diff
@@ -0,0 +1,229 @@
+
+```{r}
+#| label: combine-dfs
+gap_combined <- gap_dfs |>
+  list_rbind()
+```
+
+Now let's say that, for whatever reason, you'd like to save the data from these sheets as individual CSVs.
+This is where `walk()` comes into play---writing out the file with `write_csv()` is a "side effect." We'll use [`fs::dir_create()`](https://fs.r-lib.org/reference/create.html) to create a data folder to put our files into[^2], and build a vector of paths/file names.
+Since we have two arguments, the list of data frames, and the paths, we'll use `walk2()`.
+
+```{r}
+#| label: write-csvs
+fs::dir_create("data")
+paths <- str_glue("data/gapminder_{tolower(sheets)}.csv")
+walk2(gap_dfs, paths, write_csv)
+```
+
+To see what we've done, we can use [`fs::dir_tree()`](https://fs.r-lib.org/reference/dir_tree.html) to see the contents of the directory as a tree, or [`fs::dir_ls()`](https://fs.r-lib.org/reference/dir_ls.html) to return the paths as a list.
+
+```{r}
+#| label: fs-view
+fs::dir_tree("data")
+fs::dir_ls("data")
+```
+
+If you're having regrets, or want to return your example project to its previous state, it's just as easy to `walk()` `fs::file_delete()` along those same paths.[^3]
+
+```{r}
+#| label: fs-delete
+walk(paths, fs::file_delete)
+```
+
+## Saving multiple plots
+
+Now, let's say you want to create and save a bunch of plots.
+We'll use a modified version of the [`conditional_bars()`](https://r4ds.hadley.nz/functions.html#combining-with-other-tidyverse)[^4] function from the R for Data Science chapter on writing [functions](https://r4ds.hadley.nz/functions.html), and the built-in [diamonds](https://ggplot2.tidyverse.org/reference/diamonds.html) dataset.
+
+```{r}
+#| label: conditional-bars
+# modified conditional bars function from R4DS
+conditional_bars <- function(df, condition, var) {
+  df |>
+    filter({{ condition }}) |>
+    ggplot(aes(x = {{ var }})) +
+    geom_bar() +
+    ggtitle(rlang::englue("Count of diamonds by {{var}} where {{condition}}"))
+}
+```
+
+It's easy enough to run this for one condition, for example for the diamonds with `cut == "Good"`.
+
+```{r}
+#| label: goodclarity
+#| fig-alt: "Bar chart showing count of diamonds by clarity in the diamonds dataset where the cut == Good."
+diamonds |> conditional_bars(cut == "Good", clarity)
+```
+
+But what if we want to make a and save plot for each cut?
+Again, it's `map()` and `walk()` to the rescue.
+
+Because we're using the same data (`diamonds`) and conditioning on the same variable (`cut`), we'll only need to `map()` across the levels of `cut`, and can hard code the rest into the anonymous function.
+
+```{r}
+#| label: map-plots
+# get the levels
+cuts <- levels(diamonds$cut)
+
+# make the plots
+plots <- map(
+  cuts, 
+  \(x) conditional_bars(df = diamonds, cut == {{ x }}, clarity)
+  )
+```
+
```

Comment:
I think it'd be worth a note commenting that it's just fine to store ggplots in a list.